:Yankitute now uses :s with sub-replace

PeterRincker opened the pull request:
Use `:s` with a sub-replace gives the following benefits:
- more `:s_flags` can be handled e.g. `i` and `c`
- respects `'gdefault'`
- pattern can use `\zs` and `\ze`
- replacement can use sub-replacement aka `\=`
- provide error detection

Commit:
:Yankitute now uses :s with sub-replace

Use :s with a sub-replace gives the following benefits:
+ more :s_flags can be handled e.g. i and c
+ respects 'gdefault'
+ pattern can use \zs and \ze
+ replacement can use sub-replacement aka \=
+ provide error detection

idanarye commented:
The idea seems nice, but sadly the `\=` trick doesn't work on Vim 7.3, and many people are still using it(and open bugs on my plugins when I do something that doesn't work on it...)

Commit:
Add support for Vim version less than 7.4

+ Only use n flag when using 7.4 or 7.3 with patch
+ Do not rely on setreg able to accept a list

PeterRincker commented:
Vim 7.3 supports sub-replace-expressions. The 2 issues are as follows:
- The expression does not get evaluated when using the `n` flag.
- `setreg()` does not accept a List parameter until 7.4

I have fixed these issues and I have tested this successfully on 7.3 (the default Vim on OS X 10.10).

Commit:
Remove accidental text

idanarye commented:
Why do you need the `n` flag though? It seems to work just fine without that flag in both versions of Vim - or am I missing something?

PeterRincker commented:
The main idea is use gather the matches with the following form `[range]:s/{pat}/\=.../n{flags}`. The sub-replace-expression will capture the matches and because of the `n` flag no changes will occur to the buffer. At least that is the case for 7.3.627+.

For Vim versions _without_ the 7.3.627 patch the expression will not be ran with the `n` flag. So in this case we must filter out the `n` flag in order to run the sub-replace-expression which gathers the matches, `[range]:s/{pat}/\=.../{filtered-flags}`. Since we are no longer using the `n` flag we have the following problems:
1. Have to do a replacement, so use `\0` to make it seem like no changes occurred.
2. The cursor will now jump to the last substitution. We compensate for this with `:keepjumps` and `winsaveview()`/`winrestview()`.
3. Since we did an actually substitution (although it is virtually a no-op) the buffer is marked as modified. I do not know a good workaround for this.

Now `:Yankitute` runs on both versions of Vim.

idanarye commented:
OK, I've merged it, but I really prefer a single code path that works on all versions, so I'm going to change it to copy all matching lines to a temporary buffer and do all manipulations there, where `n` is never needed.

I'm also adding you as a collaborator, since now this plugin has more of your code than mine
